src/primitives/compression.cpp

```diff
@@ -0,0 +1,281 @@
+            m_scriptPubKey.resize(txout.scriptPubKey.size());
+            copy(txout.scriptPubKey.begin(), txout.scriptPubKey.end(), m_scriptPubKey.begin());
+    }
+}
+
+std::string CCompressedTxOut::ToString() const
+{
+    return strprintf("CCompressedTxOut(scriptPubKey=%s, scriptType=%s, nValue=%u)",
+    HexStr(m_scriptPubKey),
+    GetTxnOutputType(m_scriptType),
+    m_nValue);
+}
+
+CCompressedTransaction::CCompressedTransaction(const CTransaction tx, const uint32_t& minimumHeight, const std::vector<CCompressedOutPoint>& outs, const std::vector<CScript>& script_pubkeys) {
+    m_minimumHeight = minimumHeight;
+    m_nVersion = tx.nVersion;
+    m_nLockTime = tx.nLockTime ? tx.nLockTime-minimumHeight : 0;
+
+    for (auto const& txout : tx.vout) {
+        m_vout.push_back(CCompressedTxOut(txout));
+    }
+
+    for (size_t index = 0; index < tx.vin.size(); index++) {
+        m_vin.push_back(CCompressedTxIn(tx.vin.at(index), outs.at(index), script_pubkeys.at(index)));
+    }
+}
+
+std::string CCompressedTransaction::ToString() const
+{
+  std::string str;
+    str += strprintf("CCompressedTransaction(nVersion=%u, nLockTime=%u, minimumHeight=%u)",
+    m_nVersion,
+    m_nLockTime,
+    m_minimumHeight);
+    for (const auto& txin : m_vin)
+        str += "    " + txin.ToString() + "\n";
+    for (const auto& txout : m_vout)
+        str += "    " + txout.ToString() + "\n";
+    return str;
+}
+
+CTxOut::CTxOut(const CCompressedTxOut& txout) {
+    if (txout.IsCompressed()) {
+        if (txout.scriptType() == TxoutType::PUBKEY) {
+            scriptPubKey << txout.scriptPubKey();
+            scriptPubKey << opcodetype::OP_CHECKSIG;
+        } else {
+            std::vector<std::vector<unsigned char>> vSolutions;
+            vSolutions.push_back(std::move(txout.scriptPubKey()));
+            CTxDestination destination;
+            if(!BuildDestination(vSolutions, txout.scriptType(), destination)) throw std::ios_base::failure("Script Deseralization Failed, Either Invalid Payload or Script Type.");
+            scriptPubKey = GetScriptForDestination(destination);
+        }
+    } else {
+        scriptPubKey = CScript(txout.scriptPubKey().begin(), txout.scriptPubKey().end());
+    }
+    nValue = txout.nValue();
+}
+
+CMutableTransaction::CMutableTransaction(const CCompressedTransaction& tx, const std::vector<COutPoint>& prevouts, const std::vector<CTxOut>& outs) {
+    assert(outs.size() == tx.vin().size());
+    /* Set Values */
+    nVersion = tx.nVersion();
+    nLockTime = tx.nLockTime();
+    if (tx.nLockTime()) nLockTime += tx.minimumHeight();
+    for (const auto& txout : tx.vout()) {
+        vout.push_back(CTxOut(txout));
+    }
+    for (size_t index = 0; index < tx.vin().size(); index++) {
+        vin.push_back(CTxIn(prevouts[index], CScript(), tx.vin().at(index).nSequence()));
+    }
+
+    /* Recover Signatures */
+    for (size_t index = 0; index < tx.vin().size(); index++) {
```

Comment:
In 86c21e030c2dda625fcc3641266f8eb0fec27173:

It would be clearer to define a couple references at the top of this loop, like `CTxIn& input = tx.vin().at(index)` rather than repeating `tx.vin().at(index)` over and over, which gets noisy and is easy to confuse with `outs.at(index)` etc.

Reply:
Fixed, Thanks!